Update schtask_as.py to help bypass detection

Kahvi-0 opened the pull request:
## Description

During an engagement, the static XML file was getting detected. I found that swapping the order of settings got past detection. 

I have added some areas that shuffle the existing settings while trying to keep the format acceptable. 

![Screenshot from 2025-06-04 18-13-52](https://github.com/user-attachments/assets/74cd4083-2966-40c4-96cb-28d6c0be7893)

Note: will not print the XML file, this was just for demo.
![Screenshot from 2025-06-04 18-18-01](https://github.com/user-attachments/assets/df082fc9-b84c-427e-9d48-f8ae37becb12)


Future improvements:
- Possibly add fluff lines in there to mix things up a bit
- Setting to change the execution method from "cmd.exe" to "powershell.exe" (or w.e. else works). 

## Type of change
- [X] New feature (non-breaking change which adds functionality)


## Checklist:

- [X] I have ran Ruff against my changes (via poetry: `poetry run python -m ruff check . --preview`, use `--fix` to automatically fix what it can)
- [ ] I have added or updated the tests/e2e_commands.txt file if necessary
- [ ] New and existing e2e tests pass locally with my changes
- [ ] If reliant on changes of third party dependencies, such as Impacket, dploot, lsassy, etc, I have linked the relevant PRs in those projects
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (PR here: https://github.com/Pennyw0rth/NetExec-Wiki)

NeffIsBack commented:
> Man I love that!! @NeffIsBack  may be we should replicate that behaviour for the --exec-method atexec ?

Oh yes absolutely! @Kahvi-0 could you add that there as well?

Dfte commented:
Let's say you want to upload the LSASS_dump.exe binary using this command:

```
CMD="cmd.exe /k LSASS_dump.exe" BINARY=LSASS_dump.exe
```

Here we first upload the LSASS_dump.exe binary then we execute schtask > cmd > LSASS_dump.exe, what I wanted to do is to only have schtask > LSASS_dump.exe. Ultimately I wanted to remove the hardcoded cmd.exe /k (or powershell /k) and let the user specify the binary he wants to call, for example:

```
CMD="c:\\windows\system32\LSASS_dump.exe" BINARY=LSASS_dump.exe
```

Which will:
- Upload the LSASS_dump.exe binary
- Run the binary

This removes an unecessary process fork that could be used as an IOC.

The second reason I wanted to do such thing is that when using the actual schtask_as module, since cmd.exe is called, an actual CMD will popup on the targeted user screen which kinda sucks during """"""stealth"""""" operation (not redteam of course)

Kahvi-0 commented:
> Let's say you want to upload and run the LSASS_dump.exe binary using this command:
> 
> ```
> CMD="cmd.exe /k LSASS_dump.exe" BINARY=LSASS_dump.exe
> ```
> 
> Here we first upload the LSASS_dump.exe binary then we execute schtask > cmd > LSASS_dump.exe, what I wanted to do is to only have schtask > LSASS_dump.exe. Ultimately I wanted to remove the hardcoded cmd.exe /k (or powershell /k) and let the user specify the binary he wants to call, for example:
> 
> ```
> CMD="c:\\windows\system32\LSASS_dump.exe" BINARY=LSASS_dump.exe
> ```
> 
> Which will:
> 
> * Upload the LSASS_dump.exe binary
> * Run the binary
> 
> This removes an unecessary process fork that could be used as an IOC.
> 
> The second reason I wanted to do such thing is that when using the actual schtask_as module, since cmd.exe is called, an actual CMD will popup on the targeted user screen which kinda sucks during """"""stealth"""""" operation (not redteam of course)

I have given it a shot. From my testing it does not seem like I was able to decouple the requirement for the <exec> and <arguments> tag and let them run all on one line.

![1](https://github.com/user-attachments/assets/3987ced4-24e9-4aad-953f-202f192c87ae)

However, I believe I got a solution running. Here are some scenarios from my testing. 

Calling an existing nanodump on the system
![2](https://github.com/user-attachments/assets/6228913f-b523-42ac-a789-cf77214b94ac)

Running a CMD command
![5](https://github.com/user-attachments/assets/dd3b9cf2-9ade-4b81-aec4-093ce3927272)

Uploading nanodump.exe, running it, then deleting it
![6](https://github.com/user-attachments/assets/920d4a44-d5a4-4a0c-b62f-0654cf99e255)

Let me know if this is close to what you had in mind :)

Dfte commented:
Mate that looks perfect to me! I might take a look at it tomorrow or next week :)!
Thanks for the work!!

Kahvi-0 commented:
> Hey @Kahvi-0 so I reworked the all thing so that it only relies on a single TSCH_EXEC class #818. This is probably where we will push your upgrades as well :)!

Sounds good !

Dfte commented:
Hey @Kahvi-0, the rework was just merged https://github.com/Pennyw0rth/NetExec/pull/818. If it's okay for you, can you add your modifications ? Then I'll review the all thing this week :)!

Thanks again!

NeffIsBack commented:
It seems like there is something wrong with the commit history. The changes are applied but github still detects conflicting changes. Also it seems like main has not been merged into this branch. Did you do a rebase or something like that?

NeffIsBack commented on nxc/modules/schtask_as.py:
```diff
@@ -51,6 +50,7 @@ def options(self, context, module_options):
     name = "schtask_as"
     description = "Remotely execute a scheduled task as a logged on user"
     supported_protocols = ["smb"]
+    multiple_hosts = False
```
This has been removed in a PR a bit ago.

NeffIsBack reviewed (changes requested):
There is still something wrong with the commit history. There are changes that are reintroduced with these commits. I think manually copying out your changes and applying them to a clean branch branched out of main would be best to avoid the broken commit history

Kahvi-0 commented:
> There is still something wrong with the commit history. There are changes that are reintroduced with these commits. I think manually copying out your changes and applying them to a clean branch branched out of main would be best to avoid the broken commit history

Sorry about that, I re-pulled the latest versions and compared my changes to that and made sure the only changes I am pushing were related to this.

NeffIsBack commented on nxc/protocols/smb/atexec.py:
```diff
@@ -80,52 +82,91 @@ def get_end_boundary(self):
+        randomized_IdleSettings = "\n".join(IdleSettings)
+        random_digit = random.randint(2, 6)
+
+        match = re.match(r'^(.+?\\[^\\ ]+)\s+(.*)', command)
+        if match:
+            cmd_path = match.group(1)
+            cmd_args = match.group(2)
```
That requires now the full path to CMD.exe right? Please make that available by default.

Kahvi-0 replied:
Added a default if a user were to enter CMD=ipconfig , it will default to using CMD and also show a message to the user. 
 
<img width="672" height="474" alt="image" src="https://github.com/user-attachments/assets/02a03cca-e1a0-4979-966c-edd437492e37" />

NeffIsBack replied:
Hmm i am still not convinced. This regex will pretty sure break if you supply more complex commands or even whole powershell scripts. 

I don't know the solution at the moment, but this will likely not work. Keep in mind this is used for command execution and should not (primarily) fine tuned for schtask_as stuff. This must work with any cmd/powershell command you throw at it.

NeffIsBack commented on nxc/protocols/smb/atexec.py:
```diff
@@ -80,52 +82,91 @@ def get_end_boundary(self):
+            cmd_path = match.group(1)
+            cmd_args = match.group(2)
+        else:
+            print("Could not split the command properly.")
```
Don't use print, but the build in logger

NeffIsBack commented on nxc/protocols/smb/atexec.py:
```diff
@@ -182,40 +223,44 @@ def execute_handler(self, command):
-                smbConnection.deleteFile(self.__share, self.__output_filename)
-            except Exception:
-                pass
+            if cmdstdout == 1:
```
Why introducing yet another variable to toggle if the output should be returned? There is already `self.__retOutput`

Kahvi-0 replied:
Changed logic in both files to utilise "self.__retOutput" instead.

Kahvi-0 commented on nxc/protocols/smb/atexec.py:
```diff
@@ -80,52 +82,91 @@ def get_end_boundary(self):
+        
         xml = f"""<?xml version="1.0" encoding="UTF-16"?>
-        <Task version="1.2" xmlns="http://schemas.microsoft.com/windows/2004/02/mit/task">
+        <Task version="1.{random_digit}" xmlns="http://schemas.microsoft.com/windows/2004/02/mit/task">
```
I have the random digit choose between 2 and 6 as those were all the ones I tested that worked. I can remove this if you really think its not worth adding , but I have tested them all and worked in my lab.

NeffIsBack replied:
I exported the 2 scheduled tasks that were created automatically/via the API/UI. Both are on version 1.3. I think we should probably leave it at 1.3, because >1.3 is probably not valid and would be an easy IOC for any AV software.

Kahvi-0 commented on nxc/protocols/smb/atexec.py:
```diff
@@ -80,52 +82,91 @@ def get_end_boundary(self):
         return end_boundary.strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3]
 
     def gen_xml(self, command):
+        global cmdstdout
+        global cmd_path
```
Removed the global variables

NeffIsBack commented on nxc/protocols/smb/atexec.py:
```diff
@@ -80,52 +82,85 @@ def get_end_boundary(self):
 
-        elif self.__retOutput is False:
-            argument_xml = f"      <Arguments>/C {command}</Arguments>"
+            cmd_output = f"&gt; {self.__output_filename} 2&gt;&amp;1"
```
This just adds unnecessary complexity, please revert

NeffIsBack replied:
If you added this to increase readability just add a comment instead of nesting strings. The latter makes it harder to read

Kahvi-0 replied:
fair, reverted

NeffIsBack commented on nxc/protocols/smb/atexec.py:
```diff
@@ -192,8 +226,8 @@ def execute_handler(self, command):
                     if "STATUS_BAD_NETWORK_NAME" in str(e):
-                        self.logger.fail(f"ATEXEC: Getting the output file failed - target has blocked access to the share: {self.__share} (but the command may have executed!)")
-                        break
+                           self.logger.fail(f"ATEXEC: Getting the output file failed - target has blocked access to the share: {self.__share} (but the command may have executed!)")
+                           break
```
likely indentation issue

NeffIsBack commented on nxc/protocols/smb/atexec.py:
```diff
@@ -211,11 +245,13 @@ def execute_handler(self, command):
                         tries += 1
                         sleep(1)
-
+ 
```
Formatting

NeffIsBack commented on nxc/protocols/smb/atexec.py:
```diff
@@ -175,7 +209,7 @@ def execute_handler(self, command):
         self.logger.info(f"Deleting task \\{self.task_name}")
         tsch.hSchRpcDelete(dce, f"\\{self.task_name}")
-
+        
```
Formatting

NeffIsBack commented on nxc/protocols/smb/atexec.py:
```diff
@@ -211,11 +245,13 @@ def execute_handler(self, command):
             except Exception:
                 pass
 
+        else:
+          self.logger.display("No output file was saved to be retrived") 
```
I think that is a bit too much. If modules and or internal functions want to execute commands without any output we should not display anything in the execution class either.

NeffIsBack reviewed (changes requested):
If you use VSCode i would highly recommend to install the ruff extension, so that linting problems are displayed while working on the code. If you use something else that has no build in extensions for ruff the just run `poetry run ruff check .` before commiting&pushing changes.

Kahvi-0 commented on nxc/modules/schtask_as.py:
```diff
@@ -104,7 +102,11 @@ def on_admin_login(self, context, connection):
 
             self.logger.display(f"Executing {self.command_to_run} as {self.run_task_as}")
-            output = exec_method.execute(self.command_to_run, True)
+            #output = exec_method.execute(self.command_to_run, True)
+            if "cmd" in self.command_to_run.lower() or "powershell" in self.command_to_run.lower():
+                output = exec_method.execute(self.command_to_run, True)
+            else:
+                output = exec_method.execute(self.command_to_run, False)
```
I have removed some of the changes related to submitting a full binary path. The problem I was having before was without the base of powershell or cmd to run the binary from , redirecting output was not working. 
But setting it so the binary runs from CMD we can redirect output :

<img width="877" height="91" alt="image" src="https://github.com/user-attachments/assets/e7a5b6ff-c2c3-4708-8f11-fddf1e607952" />


<img width="970" height="425" alt="image" src="https://github.com/user-attachments/assets/74efffaf-989d-4149-958b-9497293b0431" />

Maybe we can return later to a way to supply full custom paths without the possibly of breaking things. 
I was thinking of a module option or something to control that like  "CUSTOM" that could then signal to the script to take the value of CMD as the full binary path like how I had it. 

Knowing how things go, this will pop into our heads as soon as this gets finished :p

Kahvi-0 commented:
> If you use VSCode i would highly recommend to install the ruff extension, so that linting problems are displayed while working on the code. If you use something else that has no build in extensions for ruff the just run `poetry run ruff check .` before commiting&pushing changes.

Thanks, will do that from now on. 
<img width="968" height="135" alt="image" src="https://github.com/user-attachments/assets/f84a2843-5fa0-4ddb-a8fd-77116a1976f7" />

Fixed the items you identified, main one being reverting the logic to allow full binary path for CMD, I have changed this to "C:\Windows\System32\cmd" instead of the old "cmd.exe". Wont make a huge difference with regards to detection I imagine but I still wanted to change it up somehow. This will also allow the ability to output binary output. 


For executing things under powershell / RAW paths like I had it temporarily , I was thinking of a module switch like PWSH or CUSTOM that would modify the cmd_path variable from hard coded "C:\Windows\System32\cmd" to powershell "C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe" or custom path such as "\windows\temp\rubeus.exe". I figured Id run it by you here before going for it for fear of this being non-optimal or just leaving that for a separate pull req. 

**schtask_as Examples:** 

Uploading and running binaries 
<img width="1505" height="297" alt="image" src="https://github.com/user-attachments/assets/c235e8aa-e638-4ab4-a3c7-3e735b9265b3" />
<img width="1427" height="387" alt="image" src="https://github.com/user-attachments/assets/b2e88c9d-ee36-4be5-8cec-48f5d0190482" />

Running CMD commands 
<img width="703" height="171" alt="image" src="https://github.com/user-attachments/assets/fa0c65d6-abeb-4d14-9926-75f24712311d" />


Running powershell commands (ATM formatted as:  "C:\Windows\System32\cmd /c powershell -c [...]") 
<img width="903" height="165" alt="image" src="https://github.com/user-attachments/assets/674b66cd-d721-48be-8c4d-932883459c6c" />


**atexec Examples**
powershell
<img width="550" height="118" alt="image" src="https://github.com/user-attachments/assets/838b0223-a387-43c5-a2e5-c63a8389de86" />

CMD
<img width="485" height="122" alt="image" src="https://github.com/user-attachments/assets/5c6e314f-61b9-496f-9cdd-6c0e2863986a" />

NeffIsBack commented:
> > > If you use VSCode i would highly recommend to install the ruff extension, so that linting problems are displayed while working on the code. If you use something else that has no build in extensions for ruff the just run `poetry run ruff check .` before commiting&pushing changes.
> > 
> > 
> > Thanks, will do that from now on. <img alt="image" width="968" height="135" src="https://private-user-
> 
> I have updated the PR because the latest linting changes aren't in there yet. Don't forget to pull before committing new changes.

Never mind, the PR is up-to-date, but there are still 10 linting issues. Try to execute `poetry run ruff check .`. Would anyway recommend to use poetry instead of pipx if you develop for nxc. [Here](https://www.netexec.wiki/getting-started/installation/installation-on-unix#installation-for-development-using-poetry) is the setup guide.

NeffIsBack reviewed (approved):
LGTM:
<img width="1576" height="431" alt="image" src="https://github.com/user-attachments/assets/da99e9c7-dc27-4b5f-991e-f9d0d9e0124e" />
